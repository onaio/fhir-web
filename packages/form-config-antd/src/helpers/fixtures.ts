/* eslint-disable @typescript-eslint/camelcase */
export const manifestRelease4 = {
  identifier: '1.0.30',
  json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
  appId: 'org.smartregister.foo',
  appVersion: '1.2.18',
  createdAt: '2020-06-09T14:40:04.189+02:00',
  updatedAt: '2020-06-09T14:40:04.189+02:00',
};

export const fixManifestReleases = [
  {
    identifier: '1.0.11',
    json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.14',
    createdAt: '2020-06-08T15:00:38.413+02:00',
    updatedAt: '2020-06-08T15:00:38.413+02:00',
  },
  {
    identifier: '1.0.12',
    json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.14',
    createdAt: '2020-06-08T15:03:00.950+02:00',
    updatedAt: '2020-06-08T15:03:00.950+02:00',
  },
  {
    identifier: '1.0.16',
    json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.16',
    createdAt: '2020-06-09T14:40:04.189+02:00',
    updatedAt: '2020-06-09T14:40:04.189+02:00',
  },
];

export const manifestFile1 = {
  createdAt: 'Jun 19, 2020, 12:31:37 PM',
  form_relation: '',
  id: '52',
  identifier: 'test-form-1.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'test form',
  module: '',
  version: '1.0.26',
};

export const manifestFile2 = {
  createdAt: 'Jun 19, 2020, 4:23:22 PM',
  form_relation: '',
  id: '53',
  identifier: 'reveal-test-file.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'test publish',
  module: '',
  version: '1.0.27',
};

export const manifestFile3 = {
  createdAt: 'Jun 19, 2020, 6:23:22 PM',
  form_relation: '',
  id: '54',
  identifier: 'foo.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'foo',
  module: 'baz',
  version: '1.0.28',
};

export const draftFile1 = {
  ...manifestFile1,
  isDraft: true,
  isJsonValidator: false,
};

export const draftFile2 = {
  ...manifestFile2,
  isDraft: true,
  isJsonValidator: false,
};

export const draftFile3 = {
  ...manifestFile3,
  isDraft: true,
  isJsonValidator: false,
};

export const fixManifestFiles = [manifestFile1, manifestFile2];

export const FixManifestDraftFiles = [
  {
    ...fixManifestFiles[0],
    isDraft: false,
  },
  {
    ...fixManifestFiles[1],
    isDraft: false,
  },
];

export const downloadFile = {
  clientForm: {
    json: JSON.stringify(fixManifestFiles[0]),
  },
};
