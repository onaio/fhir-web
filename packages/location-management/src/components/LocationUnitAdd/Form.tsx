import * as Yup from 'yup';
import React, { useEffect, useState } from 'react';
import { SubmitButton, Form as AntForm, Input, Radio, Select } from 'formik-antd';
import { notification, Button } from 'antd';
import { MinusCircleOutlined, PlusOutlined } from '@ant-design/icons';
import { history } from '@onaio/connected-reducer-registry';
import { getUser } from '@onaio/session-reducer';
import { OpenSRPService } from '@opensrp/server-service';
import { getAccessToken } from '@onaio/session-reducer';
import { Formik, FieldArray } from 'formik';
import { Ripple } from '@onaio/loaders';
import {
  LocationUnitPayloadPOST,
  LocationUnitPayloadPUT,
  LocationUnitStatus,
  LocationUnitSyncStatus,
} from '../../ducks/location-units';
import { useDispatch, useSelector } from 'react-redux';
import { Geometry } from 'geojson';
import { API_BASE_URL, LOCATION_TAG_ALL, LOCATION_UNIT_POST_PUT } from '../../constants';
import {
  fetchLocationTags,
  // getLocationTagsArray,
  LocationTag,
} from '../../ducks/location-tags';

const layout = { labelCol: { span: 8 }, wrapperCol: { span: 11 } };
const offsetLayout = { wrapperCol: { offset: 8, span: 11 } };

const status = [
  { label: 'Active', value: LocationUnitStatus.ACTIVE },
  { label: 'Inactive', value: LocationUnitStatus.INACTIVE },
];

// TODO : need to resolve this data from server
const parentId = [{ name: 'Bombali', value: '1123' }];

interface FormField {
  parentId: string;
  name: string;
  status: LocationUnitStatus;
  Type: string;
  externalId?: string;
  locationTags?: LocationTag[];
  geometry?: Geometry;
  textEntry?: string[];
}

const initialValue: FormField = {
  parentId: '',
  name: '',
  status: LocationUnitStatus.ACTIVE,
  Type: '',
};

/** yup validations for practitioner data object from form */
export const userSchema = Yup.object().shape({
  parentId: Yup.number().typeError('Parentid must be a Number').required('Parentid is Required'),
  name: Yup.string().typeError('Name must be a String').required('Name is Required'),
  status: Yup.string().required('Status is Required'),
  Type: Yup.string().typeError('Type must be a String').required('Type is Required'),
  externalId: Yup.string().typeError('External id must be a String'),
  locationTags: Yup.array().typeError('location Tags must be an Array'),
  geometry: Yup.string().typeError('location Tags must be a An String'),
  textEntry: Yup.array().typeError('Text Entry must be an Array'),
});

interface Props {
  id?: string;
}

export const Form: React.FC<Props> = (props: Props) => {
  const user = useSelector((state) => getUser(state));
  const accessToken = useSelector((state) => getAccessToken(state) as string);
  // const loactiontag = useSelector((state) => getLocationTagsArray(state));
  const [loactiontag, setLoactiontag] = useState<LocationTag[] | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const dispatch = useDispatch();

  useEffect(() => {
    if (isLoading) {
      let serve = new OpenSRPService(accessToken, API_BASE_URL, LOCATION_TAG_ALL);
      serve
        .list()
        .then((response: LocationTag[]) => {
          dispatch(fetchLocationTags(response));
          setIsLoading(false);
          setLoactiontag(response);
          console.log(loactiontag);
        })
        .catch((e) => console.log(e));
    }
  }, []);

  function filter(input: string, option: any) {
    return option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0;
  }

  /**
   * Handle form submission
   *
   * @param {Object} values the form fields
   * @param {Function} setSubmitting method to set submission status
   */
  function onSubmit(values: FormField, setSubmitting: (isSubmitting: boolean) => void) {
    const serve = new OpenSRPService(accessToken, API_BASE_URL, LOCATION_UNIT_POST_PUT);

    let payload: LocationUnitPayloadPOST | LocationUnitPayloadPUT = {
      properties: {
        // created: new Date(),
        // lastUpdated: new Date(),
        username: user.username,
        version: 0,
        externalId: values.externalId,
        OpenMRS_Id: props.id,
        parentId: values.parentId,
        name: values.name,
        name_en: values.name,
        status: values.status,
      },
      id: props.id,
      syncStatus: LocationUnitSyncStatus.SYNCED,
      type: values.Type,
      locationTags: values.locationTags,
      geometry: values.geometry as Geometry,
      textEntry: values.textEntry,
    };

    function removeEmptykeys(obj: any) {
      Object.keys(obj).forEach(function (key) {
        if (obj[key] && typeof obj[key] === 'object') removeEmptykeys(obj[key]);
        else if (obj[key] == null) delete obj[key];
      });
    }
    removeEmptykeys(payload);

    console.log('payload :', payload);

    if (props.id) {
      serve
        .create(payload)
        .then(() => {
          notification.success({ message: 'User created successfully', description: '' });
          setSubmitting(false);
          history.goBack();
        })
        .catch((e: Error) => {
          notification.error({ message: `${e}`, description: '' });
          setSubmitting(false);
        });
    } else {
      serve
        .update(payload)
        .then(() => {
          notification.success({ message: 'User Updated successfully', description: '' });
          setSubmitting(false);
          history.goBack();
        })
        .catch((e: Error) => {
          notification.error({ message: `${e}`, description: '' });
          setSubmitting(false);
        });
    }
  }

  if (isLoading) return <Ripple />;

  return (
    <Formik
      initialValues={initialValue}
      validationSchema={userSchema}
      onSubmit={(
        values: FormField,
        { setSubmitting }: { setSubmitting: (isSubmitting: boolean) => void }
      ) => onSubmit(values, setSubmitting)}
    >
      {({ values, errors, isSubmitting, handleSubmit }) => {
        console.log('values :', values, ' errors :', errors);

        return (
          <AntForm requiredMark={'optional'} {...layout} onSubmitCapture={handleSubmit}>
            <AntForm.Item label="Parent" name="parentId" required>
              <Select
                name="parentId"
                showSearch
                placeholder="Select a Parent Id"
                optionFilterProp="children"
                filterOption={filter}
              >
                {parentId.map((e) => (
                  <Select.Option key={e.value} value={e.value}>
                    {e.name}
                  </Select.Option>
                ))}
              </Select>
            </AntForm.Item>

            <AntForm.Item name="name" label="Name" required>
              <Input name="name" placeholder="Enter a location name" />
            </AntForm.Item>

            <AntForm.Item label="Status" name="status" valuePropName="checked" required>
              <Radio.Group name="status" defaultValue={initialValue.status}>
                {status.map((e) => (
                  <Radio name="status" key={e.label} value={e.value}>
                    {e.label}
                  </Radio>
                ))}
              </Radio.Group>
            </AntForm.Item>

            <AntForm.Item name="Type" label="Type" required>
              <Input name="Type" placeholder="Select type" />
            </AntForm.Item>

            <AntForm.Item name="externalId" label="External ID">
              <Input name="externalId" placeholder="Select status" />
            </AntForm.Item>

            <AntForm.Item name="geometry" label="geometry">
              <Input.TextArea name="geometry" rows={4} placeholder="</> JSON" />
            </AntForm.Item>

            <AntForm.Item label="Unit Group" name="locationTags">
              <Select
                name="locationTags"
                mode="multiple"
                allowClear
                showSearch
                placeholder="Enter a location group name"
                optionFilterProp="children"
                filterOption={filter}
              >
                {loactiontag &&
                  loactiontag.map((e) => (
                    <Select.Option key={e.id} value={e.id}>
                      {e.name}
                    </Select.Option>
                  ))}
              </Select>
            </AntForm.Item>

            <FieldArray
              name="textEntry"
              render={(arrayHelpers) => (
                <>
                  {values.textEntry &&
                    values.textEntry.length > 0 &&
                    values.textEntry.map((field, index) => {
                      const key = `textEntry.${index}`;
                      return (
                        <AntForm.Item
                          name={key}
                          {...(index === 0
                            ? { labelCol: { span: 8 }, wrapperCol: { span: 16 } }
                            : { wrapperCol: { offset: 8, span: 16 } })}
                          label={index === 0 ? 'Text entry' : ''}
                          key={key}
                        >
                          <AntForm.Item name={key} noStyle>
                            <Input name={key} placeholder="Enter text" style={{ width: '69%' }} />
                          </AntForm.Item>
                          <MinusCircleOutlined
                            className="dynamic-delete-button"
                            onClick={() => arrayHelpers.remove(index)}
                          />
                        </AntForm.Item>
                      );
                    })}

                  <AntForm.Item name="textEntryButton" {...offsetLayout}>
                    <Button
                      type="dashed"
                      onClick={() => arrayHelpers.push('')}
                      style={{ width: '100%' }}
                      icon={<PlusOutlined />}
                    >
                      Add Text Entry field
                    </Button>
                  </AntForm.Item>
                </>
              )}
            />
            <AntForm.Item name="buttons" {...offsetLayout}>
              <SubmitButton id="submit">{isSubmitting ? 'Saving' : 'Save'}</SubmitButton>
              <Button id="cancel" onClick={() => history.goBack()} type="dashed">
                Cancel
              </Button>
            </AntForm.Item>
          </AntForm>
        );
      }}
    </Formik>
  );
};

export default Form;
