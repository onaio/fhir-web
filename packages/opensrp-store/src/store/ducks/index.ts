/** users ducks modules: actions, actionCreators, reducer and selectors */
import { Dictionary } from '@onaio/utils';
import { keyBy, values } from 'lodash';
import intersect from 'fast_array_intersect';
import { AnyAction, Store } from 'redux';
import { createSelector, OutputParametricSelector } from 'reselect';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'keycloakUsers';

/** Interface for user json object */
export interface KeycloakUser {
  access?: {
    manageGroupMembership: boolean;
    view: boolean;
    mapRoles: boolean;
    impersonate: boolean;
    manage: boolean;
  };
  createdTimestamp?: number;
  disableableCredentialTypes?: string[];
  email?: string;
  emailVerified?: boolean;
  enabled?: boolean;
  firstName: string;
  id: string;
  lastName: string;
  notBefore?: number;
  requiredActions?: string[];
  totp?: boolean;
  username: string;
}

// actions

/** action type for fetching keycloak users */
export const KEYCLOAK_USERS_FETCHED = 'keycloak/reducer/users/userS_FETCHED';
/** action type for removing keycloak users */
export const REMOVE_KEYCLOAK_USERS = 'keycloak/reducer/users/REMOVE_KEYCLOAK_USERS';

/** interface action to add users to store */
export interface FetchKeycloakUsersAction extends AnyAction {
  usersById: Dictionary<KeycloakUser>;
  type: typeof KEYCLOAK_USERS_FETCHED;
}

/** Interface for removeusersAction */
export interface RemoveKeycloakUsersAction extends AnyAction {
  usersById: Dictionary<KeycloakUser>;
  type: typeof REMOVE_KEYCLOAK_USERS;
}

/** Create type for users reducer actions */
export type KeycloakUsersActionTypes =
  | FetchKeycloakUsersAction
  | RemoveKeycloakUsersAction
  | AnyAction;

// action Creators

/**
 * Fetch users action creator
 *
 * @param {Array} usersList - keycloak users
 * @returns {object} - the dispatcher to remove the user
 */
export const fetchKeycloakUsers = (usersList: KeycloakUser[] = []): FetchKeycloakUsersAction => {
  return {
    usersById: keyBy(usersList, (user: KeycloakUser) => user.id),
    type: KEYCLOAK_USERS_FETCHED,
  };
};

/** Remove users action creator
 *
 * @returns {object} - the dispatcher to remove the user
 */
export const removeKeycloakUsers = (): RemoveKeycloakUsersAction => {
  return {
    usersById: {},
    type: REMOVE_KEYCLOAK_USERS,
  };
};

// The reducer

/** interface for keycloak users state in redux store */
export interface KeycloakUsersState {
  usersById: Dictionary<KeycloakUser> | Dictionary;
}

/** Create an immutable keycloak users state */
export type ImmutableKeycloakUsersState = KeycloakUsersState &
  SeamlessImmutable.ImmutableObject<KeycloakUsersState>;

/** initial keycloak users state */
export const initialState: ImmutableKeycloakUsersState = SeamlessImmutable({
  usersById: {},
});

/**
 * the users reducer function
 *
 * @param {object} state - keycloak users states
 * @param {object} action - keycloak users actions
 * @returns {object} - the updated states
 */
export function reducer(
  state: ImmutableKeycloakUsersState = initialState,
  action: KeycloakUsersActionTypes
): ImmutableKeycloakUsersState {
  switch (action.type) {
    case KEYCLOAK_USERS_FETCHED:
      return SeamlessImmutable({
        ...state,
        usersById: { ...state.usersById, ...action.usersById },
      });
    case REMOVE_KEYCLOAK_USERS:
      return SeamlessImmutable({
        ...state,
        usersById: action.usersById,
      });

    default:
      return state;
  }
}

// Selectors

export interface KeycloakUsersFilters {
  id: string[] /** get all users whose ids appear in this array */;
  username?: string /** get users whose username includes text in the username field */;
}

// NON MEMOIZED SELECTORS

/**
 * Gets id from KeycloakUsersFilters
 *
 * @param {any} _ - the redux store
 * @param {Array} props - the keycloak users filters object
 * @returns {string} - the keycloak user id
 */
export const getUserIds = (_: Partial<Store>, props: KeycloakUsersFilters): string[] => props.id;

/**
 * Gets name from KeycloakUsersFilters
 *
 * @param {any} _ - the redux store
 * @param {object} props - the users filters object
 * @returns {string} - the keycloak user name
 */
export const getUsername = (_: Partial<Store>, props: KeycloakUsersFilters): string | undefined =>
  props.username;

/** returns all users in the store as values whose keys are their respective ids
 *
 * @param {any} state - the redux store
 * @returns {object} - users object as values, respective ids as keys
 */
export function getKeycloakUsersById(state: Partial<Store>): { [key: string]: KeycloakUser } {
  return (state as Dictionary)[reducerName].usersById;
}

/** gets keycloak users as an array of user objects
 *
 * @param {any} state - the redux store
 * @returns {Array} - an array of users objs
 */
export function getKeycloakUsersArray(state: Partial<Store>): KeycloakUser[] {
  return values(getKeycloakUsersById(state));
}

/**
 * Gets all practitioners whose name includes phrase given in name filter prop
 *
 * @returns {Array} - practitioners whose name includes phrase given in name filter prop
 */
export const getUsersByUsername = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (res1: KeycloakUser[], res2: string | undefined) => KeycloakUser[]
> =>
  createSelector(getKeycloakUsersArray, getUsername, (usersArray, username) =>
    username
      ? usersArray.filter((user: KeycloakUser) =>
          user.username.toLowerCase().includes(username.toLowerCase())
        )
      : usersArray
  );

// MEMOIZED SELECTORS

/**
 * Gets all users whose identifiers appear in ids filter prop value
 *
 * @returns {Array} - users whose identifiers appear in ids filter prop value
 */
export const getKeycloakUsersByIds = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (
    res1: {
      [key: string]: KeycloakUser;
    },
    res2: string[],
    res3: KeycloakUser[]
  ) => KeycloakUser[]
> =>
  createSelector(
    getKeycloakUsersById,
    getUserIds,
    getKeycloakUsersArray,
    (usersById, ids, usersArray) => {
      if (ids === undefined) {
        return usersArray;
      }
      if (ids.length > 0) {
        return ids.map((id: string) => usersById[id]);
      }
      return [];
    }
  );

/** practitioner array selector factory
 * aggregates response from all applied filters and returns results
 *
 * @returns {Array} - aggregates response from all applied filters and returns results
 */
export const makeKeycloakUsersSelector = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (res1: KeycloakUser[], res2: KeycloakUser[]) => KeycloakUser[]
> =>
  createSelector(getKeycloakUsersByIds(), getUsersByUsername(), (arr1, arr2) => {
    return intersect([arr1, arr2], JSON.stringify);
  });
