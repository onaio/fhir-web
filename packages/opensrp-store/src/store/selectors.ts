import { reducerName as sessionReducerName } from '@onaio/session-reducer';
import { Store } from 'redux';
import { Dictionary } from '@onaio/utils';
import { createSelector } from 'reselect';

/** get API Token from the Redux store
 *
 * @param {any} state - the redux store
 */
interface StoreState {
  session: {
    [key: string]: Dictionary;
  };
}

interface APIFilters {
  accessToken?: boolean;
  apiToken?: boolean;
  providerState?: boolean;
}

/** Fetch access token from store if access_token filter val is true
 *
 * @param {object} _ - redux store
 * @param {object} props - api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchAccessToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.accessToken || null;

/** Fetch api token from store if api_token filter val is true
 *
 * @param {object} _ redux store
 * @param {object} props api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchApiToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.apiToken || null;

/** Fetch oauth provider state from store if providerState filter val is true
 *
 * @param {Object} _ redux store
 * @param {Object} props api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchOauthProviderState = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.providerState || null;

/** Gets extra data object from store
 *
 * @param {object} state - redux store
 * @returns {Object}
 */

export const getExtraData = (state: Partial<Store>): Dictionary =>
  (state as StoreState)[sessionReducerName].extraData;

/** API state selector
 *
 * @returns {Object} - the states
 */
export const makeAPIStateSelector = () =>
  createSelector(
    [fetchAccessToken, fetchApiToken, fetchOauthProviderState, getExtraData],
    (accessToken, apiToken, oauthState, extraData) => {
      if (accessToken) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.access_token) || null;
      } else if (apiToken) {
        return extraData.api_token || null;
      } else if (oauthState) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.state) || null;
      }
    }
  );
