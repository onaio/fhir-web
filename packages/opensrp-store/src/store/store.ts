import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';
import { Reducer } from 'redux';
import { RouterState } from 'connected-react-router';
import { Dictionary } from '@onaio/utils';

/** declare globals interface */
declare global {
  interface Window {
    __PRELOADED_STATE__: Dictionary;
  }
}

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as Reducer<RouterState>,
};

// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__ || {};
// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__;

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
export const store = getConnectedStore(defaultReducers, preloadedState);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener((reducers) => {
  store.replaceReducer(combine(reducers));
});
