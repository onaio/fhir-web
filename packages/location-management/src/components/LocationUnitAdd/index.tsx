import React from 'react';
import { Helmet } from 'react-helmet';
import reducerRegistry from '@onaio/redux-reducer-registry';
import reducer, { reducerName } from '../../ducks/location-units';
import Form from './Form';
import { useParams } from 'react-router';

reducerRegistry.register(reducerName, reducer);

export const LocationUnitAdd: React.FC = () => {
  const params: { id: string } = useParams();
  const { id } = params;
  return (
    <section>
      <Helmet>
        <title>Add Location Unit Group</title>
      </Helmet>

      <h5 className="mb-3">Add Location Unit Group</h5>

      <div className="bg-white p-5">
        <Form id={id} />
      </div>
    </section>
  );
};

export default LocationUnitAdd;
