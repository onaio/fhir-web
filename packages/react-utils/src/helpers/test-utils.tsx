import { store } from '@opensrp/store';
import { Provider } from 'react-redux';
import React from 'react';
import { RoleContext, UserRole, FhirResources, Permit, IamResources } from '@opensrp/rbac';
import { history } from '@onaio/connected-reducer-registry';
import { Router } from 'react-router';

export const superUserRole = new UserRole([...FhirResources, ...IamResources], Permit.MANAGE);

export interface ContextProviderProps {
  children: JSX.Element;
}

export const ContextProvider = (props: ContextProviderProps) => {
  const { children } = props;
  return (
    <Provider store={store}>
      <Router history={history}>
        <RoleContext.Provider value={superUserRole}>{children}</RoleContext.Provider>
      </Router>
    </Provider>
  );
};
