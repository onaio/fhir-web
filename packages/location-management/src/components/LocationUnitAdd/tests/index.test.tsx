import flushPromises from 'flush-promises';
import { mount } from 'enzyme';
import React from 'react';
import { history } from '@onaio/connected-reducer-registry';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Router } from 'react-router';
import fetch from 'jest-fetch-mock';
import { store } from '@opensrp/store';

import { id, locationtag, sampleHierarchiesList, sampleLocationUnit } from './fixtures';
import LocationUnitAdd from '..';
import { act } from 'react-dom/test-utils';
import { sampleHierarchy } from '../../LocationUnitView/tests/fixtures';
import toJson from 'enzyme-to-json';

describe('containers/pages/locations/LocationUnitAdd', () => {
  it('renders without crashing', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationUnitAdd />
        </Router>
      </Provider>
    );
    expect(wrapper.props()).toMatchSnapshot();

    await act(async () => {
      await flushPromises();
    });
    wrapper.unmount();
  });

  it('renders without crashing when editting', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ pathname: `/${id}`, hash: '', search: '', state: {} }]}>
          <Route path={'/:id'} component={LocationUnitAdd} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.props()).toMatchSnapshot();

    await act(async () => {
      await flushPromises();
    });
    wrapper.unmount();
  });

  it('works correctly with store', async () => {
    fetch
      .once(JSON.stringify(sampleLocationUnit))
      .once(JSON.stringify(locationtag))
      .once(JSON.stringify(sampleHierarchiesList))
      .once(JSON.stringify(sampleHierarchy));
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ pathname: `/${id}`, hash: '', search: '', state: {} }]}>
          <Route path={'/:id'} component={LocationUnitAdd} />
        </MemoryRouter>
      </Provider>
    );
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    const fetchCalls = [
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location-tag',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/findByProperties?is_jurisdiction=true&return_geometry=false&properties_filter=status%3AActive%2CgeographicLevel%3A0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/a26ca9c8-1441-495a-83b6-bb5df7698996?is_jurisdiction=true',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location-tag',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/findByProperties?is_jurisdiction=true&return_geometry=false&properties_filter=status%3AActive%2CgeographicLevel%3A0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/a26ca9c8-1441-495a-83b6-bb5df7698996?is_jurisdiction=true',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location-tag',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/findByProperties?is_jurisdiction=true&return_geometry=false&properties_filter=status%3AActive%2CgeographicLevel%3A0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/5a286fc9-e985-4aa7-a843-d71f93d1f4b4',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/9feb1a4d-3020-4f1a-bbd9-0a47b54d2ae9',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/b652b2f4-a95d-489b-9e28-4629746db96a',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/35bf4771-a404-4220-bd9e-e2916decc116',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/a68d8a73-f235-4b9c-9717-41ba2546d771',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/66be3e59-5d49-4051-8ea4-dff7503d8d69',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/4f12391e-48ce-40d7-9dbb-813e93d042cb',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/c692b90d-ee20-4240-8904-8cbbbcab54f7',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/a26ca9c8-1441-495a-83b6-bb5df7698996',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
    ];
    expect(fetch.mock.calls).toEqual(fetchCalls);
  });
});
