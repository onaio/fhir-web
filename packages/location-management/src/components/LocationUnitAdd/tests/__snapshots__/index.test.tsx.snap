// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/locations/LocationUnitAdd renders without crashing 1`] = `
Object {
  "children": <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <LocationUnitAdd />
  </Router>,
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
}
`;

exports[`containers/pages/locations/LocationUnitAdd renders without crashing 2`] = `
Object {
  "children": <MemoryRouter
    initialEntries={
      Array [
        Object {
          "hash": "",
          "pathname": "/a26ca9c8-1441-495a-83b6-bb5df7698996",
          "search": "",
          "state": Object {},
        },
      ]
    }
  >
    <Route
      component={[Function]}
      path="/:id"
    />
  </MemoryRouter>,
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
}
`;

exports[`containers/pages/locations/LocationUnitAdd renders without crashing when editting 1`] = `
Object {
  "children": <MemoryRouter
    initialEntries={
      Array [
        Object {
          "hash": "",
          "pathname": "/a26ca9c8-1441-495a-83b6-bb5df7698996",
          "search": "",
          "state": Object {},
        },
      ]
    }
  >
    <Route
      component={[Function]}
      path="/:id"
    />
  </MemoryRouter>,
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
}
`;
